ciongke/cmd/hook/main.go

```diff
@@ -0,0 +1,263 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"crypto/hmac"
+	"crypto/sha1"
+	"encoding/hex"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/google/go-github/github"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/resource"
+	"k8s.io/kubernetes/pkg/apis/batch"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/labels"
+)
+
+var (
+	port              = flag.Int("port", 8888, "Port to listen on.")
+	webhookSecretFile = flag.String("webhook-secret-file", "/etc/hook/token", "Path to a file containing the GitHub webhook secret.")
+)
+
+var (
+	jobDeadlineSeconds int64 = 3600
+	jobResourceCPU     int64 = 2
+	jobResourceMemory  int64 = 4 * 1000 * 1000 * 1000
+
+	// TODO: Load these from a configmap.
+	jobImage  = "gcr.io/k8s-ci-on-gke/test-pr:0.0"
+	jobBucket = "k8s-ci-on-gke-results"
+)
+
+// Server implements http.Handler.
+type Server struct {
+	Port   int
+	Secret []byte
+	Events chan Event
+}
+
+type Event struct {
+	Type    string
+	Payload []byte
+}
+
+func main() {
+	flag.Parse()
+
+	sb, err := ioutil.ReadFile(*webhookSecretFile)
+	if err != nil {
+		log.Fatalf("Could not read secret file: %s", err)
+	}
+
+	config, err := restclient.InClusterConfig()
+	if err != nil {
+		log.Fatalf("Error getting rest client: %s\n", err)
+	}
+
+	s := &Server{
+		Port:   *port,
+		Secret: sb,
+		Events: make(chan Event),
+	}
+	go func() {
+		if err := http.ListenAndServe(":"+strconv.Itoa(*port), s); err != nil {
+			log.Fatalf("ListenAndServe returned error: %s", err)
+		}
+	}()
+	for event := range s.Events {
+		go handleEvent(config, event)
+	}
```

Comment:
Minor nit: I would have this loop in a go routine and not the `ListendAndServe`.